analyzers/src/SonarAnalyzer.CSharp/Rules/MemberInitializerRedundant.cs

```diff
@@ -56,128 +56,189 @@ protected override void Initialize(SonarAnalysisContext context)
+                                                  || method.IsModuleInitializer(),
+                    // only static members are interesting (const ones can only be initialized once)
+                    FieldFilter = field => field.Modifiers.Any(IsStatic),
+                    PropertyFilter = property => property.Modifiers.Any(IsStatic),
+                    IsSymbolFirstSetInInitializers = (fieldOrPropertySymbol, initializerDeclarations) =>
+                        // there can be only one static constructor
+                        // all module initializers are executed when the type is created, so it is enough if ANY initializes the member
+                        initializerDeclarations.Any(x => IsSymbolFirstSetInCfg(fieldOrPropertySymbol, x.Node, x.SemanticModel)),
+                    DiagnosticMessage = StaticMemberMessage
+                };
+
+            private static bool IsStaticOrConst(SyntaxToken token) => token.IsKind(SyntaxKind.StaticKeyword) || token.IsKind(SyntaxKind.ConstKeyword);
+
+            private static bool IsStatic(SyntaxToken token) => token.IsKind(SyntaxKind.StaticKeyword);
+
+            /// <summary>
+            /// Returns true if the member is overwritten without being read in the instance constructor.
+            /// Returns false if the member is not set in the constructor, or if it is read before being set.
+            /// </summary>
+            private static bool IsSymbolFirstSetInCfg(ISymbol classMember, BaseMethodDeclarationSyntax constructorOrInitializer, SemanticModel semanticModel)
             {
-                return false;
-            }
+                var body = (CSharpSyntaxNode)constructorOrInitializer.Body ?? constructorOrInitializer.ExpressionBody();
+                if (!CSharpControlFlowGraph.TryGet(body, semanticModel, out var cfg))
+                {
+                    return false;
+                }
 
-            var checker = new MemberInitializerRedundancyChecker(cfg, declaredSymbol, ctor.SemanticModel);
-            return checker.CheckAllPaths();
+                var checker = new MemberInitializerRedundancyChecker(cfg, classMember, semanticModel);
+                return checker.CheckAllPaths();
+            }
         }
 
-        private static IEnumerable<DeclarationTuple<IPropertySymbol>> GetInitializedPropertyDeclarations(TypeDeclarationSyntax declaration,
-            SemanticModel semanticModel)
+        private class InitializationVerifier<TInitializerDeclaration> where TInitializerDeclaration : SyntaxNode
         {
-            return declaration.Members
-                .OfType<PropertyDeclarationSyntax>()
-                .Where(p => !p.Modifiers.Any(IsStaticOrConst) &&
-                    p.Initializer != null &&
-                    p.IsAutoProperty())
-                .Select(p =>
-                    new DeclarationTuple<IPropertySymbol>
+            public Func<IMethodSymbol, bool> InitializerFilter;
```

Comment:
I would convert the fields to private readonly and create a constructor to set the values there. That way it will be more clear that the values have to be set in order the class to do the job.